File: Stage21/Python/Problem12865.py
def main():
    N, K = map(int, input().split())
    dp = [0] * (K + 1)

    for _ in range(N):
        W, V = map(int, input().split())
        for w in range(K, W - 1, -1):
            dp[w] = max(dp[w], dp[w - W] + V)

    print(dp[K])

if __name__ == "__main__":
    main()

File: Stage21/C/Problem12865.c
#include <stdio.h>

int max(int a, int b) {
    return a > b ? a : b;
}

int main(void) {
    int N, K;
    scanf("%d %d", &N, &K);

    int dp[100001] = {0};

    for (int i = 0; i < N; i++) {
        int W, V;
        scanf("%d %d", &W, &V);

        for (int w = K; w >= W; w--) {
            dp[w] = max(dp[w], dp[w - W] + V);
        }
    }

    printf("%d\n", dp[K]);
    return 0;
}